### PR TITLE
Add system user to avoid running container with root

### PR DESCRIPTION
**Pull request by ToadCast**
I have done some deployment with this image, for security reasons I propose to add a user instead or keeping root running it.

### CODE REVIEW

**Comment by tianon** [comment]
Great catch - this is definitely a bit of an ugly oversight.  However, I have a couple issues with this PR as-is:

1. inconsistent with other docker-library images / not optimized for docker build cache -- the `useradd` bits should be moved up as high as possible, especially since the user/group added has a specific ID, and `99` is not an ID we've used anywhere else that I'm aware of; it's hard to make a choice here because no matter what we choose it's likely to break someone (see point 3 below)
2. `setcap` is very unreliable in my experience, especially round-tripping through Docker graph drivers; users who need "privileged" ports are much better off using `--sysctl net.ipv4.ip_unprivileged_port_start=0`
3. backwards compatibility - this is very likely to break many users, so I don't think we can include a `USER` instruction, but in this case users have to supply their own configuration for the image to run anyhow, so I think we can solve that by making this half a documentation change instead (and perhaps adding it for future releases / the in-progress dev releases only?)

**Comment by tianon** [comment]
I've pushed an update to this that moves the user creation up as high as we can and only sets `USER` in 2.4+ (the current pre-release / dev releases).

I also verified that UID/GID 99 appears to be reasonably unused by default in both Debian and Alpine (checking `aports` for static assignments), so it seems like a reasonably safe choice.  I'm concerned that us creating an `haproxy` user by default might cause issues for downstream users who are currently doing so, but I don't think that's really avoidable.

**Comment by tianon** [comment]
(Thanks for kicking off this conversation/change! :+1: :heart:)

**Comment by ToadCast** [comment]
Thanks for your help into this PR, I appreciate.
I'm curious about the best practice mentionning that we should put the useradd as high as possible, could you enlight me or send me somewhere I could find it? Because so far I've read about layers should be easy to process, and wherever is the useradd shouldn't be a problem.

**Comment by tianon** [comment]
I think https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#leverage-build-cache is pretty good. :slightly_smiling_face:

Concretely, moving this up means that if something lower down changes, this layer doesn't have to rebuild (and means that the different versions of `haproxy` can share this layer, so if a user ends up running more than one version at once it will require slightly less storage/transfer).

**Review by yosifkit — APPROVED** [review]

File: Dockerfile-debian.template [hunk 1]
```diff
@@ -73,6 +84,9 @@ RUN set -eux; \
 # "graceful stop is triggered when the SIGUSR1 signal is sent to the haproxy process"
 STOPSIGNAL SIGUSR1
 
-COPY docker-entrypoint.sh /
-ENTRYPOINT ["/docker-entrypoint.sh"]
+COPY docker-entrypoint.sh /usr/local/bin/
+RUN ln -s usr/local/bin/docker-entrypoint.sh / # backwards compat
+ENTRYPOINT ["docker-entrypoint.sh"]
+
+USER haproxy
```
Comment (TimWolla) [inline Dockerfile-debian.template:91]: Note: It is recommended to launch HAProxy as root and to specify `user` within the configuration instead. This allows HAProxy to adjust its ulimit (e.g. the number of file descriptors) and also to chroot itself.

Comment (tianon) [reply Dockerfile-debian.template:91]: Hmm, I guess we've got a difficult choice to make then - setting both of those via the container runtime is going to be strictly more secure (and a `chroot` inside a container is probably not as compelling).

Comment (TimWolla) [reply Dockerfile-debian.template:91]: Yeah, I understand that having the container around HAProxy changes the situation a bit. I can't recommend on the "HAProxy in Docker" perspective here. I run all services in Docker, but HAProxy is very intentionally not within a container and instead deployed on the host.

I've also just verified that having a `user haproxy` in the configuration will not require any changes to the config (a `user somethingelse` will however), so this change should also be compatible with copy and paste configurations found on the Internet (unless they also put `chroot`).

I'll leave it up to you to decide what to do with the `USER` like, I just wanted to note this from a purely HAProxy perspective :smile:

Comment (ToadCast) [reply Dockerfile-debian.template:91]: I guess that's getting tricky then. The thing is, on some platform, you can not start a container with an entrypoint that is done by `root`. EKS rejects the initial image for that reason, and I tried to have the process handled with the user, but only to start the container with `USER` could make it work.

**Pull request merged**